CenterForOpenScience/osf.io
# [Feature] Show all projects in 'link projects' [OSF-5360]

## Purpose

Show all user projects and registrations in 'link projects' or 'add node link' modal.

Currently, the user has to know the name of the project/registrations to which they wish
to link a project. With name, they can search but registrations and projects results are not distinguishable.

This PR update adds now available  API support for distinguishing node links vs registration links,
presumably (add/remove registrations links).

## Changes

Build from PR https://github.com/CenterForOpenScience/osf.io/pull/6137 from [mluke93](https://github.com/mluke93) which was closed due to lack of API support for /registration_links/

This PR update:
- Refactor code 
- Reduce requests needed to tell a node's node-links given myprojects or all public projects.
- Fix issues with loaders and use new css loaders
- Add clear/cleanup function at modal exit.
- Only reload if user adds/removes a node link
 
## Ticket

[OSF-5360](https://openscience.atlassian.net/browse/OSF-5360)

## Before

![before](https://user-images.githubusercontent.com/4511563/28578251-25f42986-7127-11e7-84ee-188f919c6348.gif)

## After
![after-responsiveness-fix-2](https://user-images.githubusercontent.com/4511563/30550811-99d5e154-9c66-11e7-90dc-f28da721e7da.gif)

**latest**
![latest-fix](https://user-images.githubusercontent.com/4511563/30609906-a27d6f2c-9d4a-11e7-8b86-9ecbfcb36c0f.gif)


## QA considerations
Since the ticket has been sitting for a while, please make sure the behavior is as shown in the latest gif. Mostly regular usage.
- Check if guid searching works
- See if the the modal behaves as expected for smaller screens and mobile
- See if users with only one name display correctly
- See if you leave the query in the search box, and try pressing the different options (myprojects, all projects, registrations, etc.) they display results related to that query.
- Make sure the current project or its current descendants are not returned as nodes you can link to.
